firmware/ventilator-controller-stm32/Core/Test/Pufferfish/Protocols/crcelements.txt

```diff
@@ -0,0 +1,104 @@
+      THEN("the computed crc is equal to the expected crc")
+
+  GIVEN("A crc element is initalised with payload 0x01")
+    WHEN("the crc is computed from the payload")
+      THEN("the computed crc is equal to the expected crc")
+
+/* write function
+Scenario: CRCElement: The write function correctly generates body from computed crc and payload
+  GIVEN("A crc element initalised with payload=[0x01, 0x02, 0x05]")
+    WHEN("The crc and payload are written to the output buffer")
+      THEN("crc computed in write method is same as crcchecker compute")
+      THEN("The output buffer is as expected")
+
+  GIVEN("A crc element is initalised with an empty payload")
+    WHEN("The crc and payload are written to the output buffer")
+      THEN("The CRC returned by the crc() method is equal to 0")
+      THEN("The output buffer is as expected")
+
+/* parse function
+Scenario: CRCElement: The parse function correctly updates internal crc and payload fields from the input buffer
+
+  GIVEN("A CRC Element initialized with an empty payload buffer with a capacity of 254 bytes")
+
+    WHEN("A body with crc more than 4 bytes and equal to the crc of the payload is parsed")
+      THEN("the parse method reports ok status")
+      THEN("crc returned from the crc() method is same as the crc of the payload in input body")
+      THEN("payload returned from the payload() method is equal to the payload from the body")
+
+    WHEN("A body of less than 4 bytes is parsed")
+      THEN("the parse method reports out of bounds status")
+
+    WHEN("A body with crc and payload is given, where the crc is not equal to the actual crc of the payload")
+      THEN("the parse method reports ok status")
+      THEN("The crc returned from the crc() method is equal to the crc in the body/input_buffer")
+      THEN("The crc returned from the crc() method is not equal to the actual crc for the given payload")
+      THEN("The payload returned from the payload() method is equal to the payload from the body")
+
+/* write parse roundtrip
+Scenario: The crc element correctly writes and parses to a buffer
+
+  GIVEN("A CRC Element initialized with a non empty payload buffer with a capacity of 254 bytes")
+
+    WHEN("A body with crc and payload is generated, parsed and generated again")
+      # write  
+      THEN("The status of write function returns ok")
+      THEN("The crc from the crc element is equal to expected")
+      THEN("The payload of the crc element is as expected")
+      THEN("The output buffer is as expected")
+      # parse
+      THEN("The status of parse function returns ok")
+      THEN("The crc from the CRCElement is equal to the crc in the body")
+      THEN("payload from the CRCElement is equal to the expected payload from the body")
+      # write  
+      THEN("The status of write function returns ok")
+      THEN("The crc from the crc element is equal to expected")
+      THEN("The payload of the crc element is as expected")
```

Comment:
Comments and suggestions made about THENs for the individual write and parse scenarios above also apply here. In general these THENs should be changed to look as similar as possible to those previous THENs.